### ISSUE TITLE
Incorrect diff value placement

### ISSUE THREAD

**Issue by hcpl** [colin-kiegel/rust-pretty-assertions#12]
### Screenshot

![2017-09-05-183035-506x464_maim_slop](https://user-images.githubusercontent.com/17436405/30069300-0b6de7a6-9269-11e7-891c-f9d6a86a0318.png)

### Code to reproduce

```rust
#[macro_use]
extern crate pretty_assertions;


#[cfg(test)]
mod tests {
    #[test]
    fn has_bug() {
        let left = vec![0, 0, 0, 128, 10, 191, 5, 64];
        let right = vec![84, 248, 45, 64];

        assert_eq!(left, right);
    }
}
```

**Comment by colin-kiegel** [comment]
@hcpl Thank you for this excellent report :-)

As part of this fix, I intend to improve the readability of character-level differences. Here are two versions of a simplified example. I find version A easier to read, but it doesn't highlight differences in whitespace characters.

# A
![issue12_a](https://user-images.githubusercontent.com/7691186/30251390-e9a7a75c-965e-11e7-8c59-ee0967aefd15.png)

# B
![issue12_b](https://user-images.githubusercontent.com/7691186/30251392-eb40cdbe-965e-11e7-8c67-90380b0d2e91.png)

**Comment by hcpl** [comment]
Dark theme:
- A is easiest to read;
- B lacks contrast but contains whitespace diff;
- C both is easy to read and contains whitespace diff.

Light theme:
- A looks awful, especially when the screen brightness is > 90% and you're looking at text at some angle;
- B is pretty good, but diff`ed characters can be too dark to spot immediately;
- C is perfect.

So overall C is best for me if I had to stick to a single one. But the colors could be configurable if this is possible.

**Comment by fflorent** [comment]
I wish that features (or something similar in `Cargo.toml`) could give some ability to pass strings to the code, so we can give to the user more control of how the diff are rendered.

Say something like:
````toml
# in Cargo.toml
left_fg_color="green"
right_fg_color="red"
````

Do you know anything that could the trick?

Florent

**Comment by hcpl** [comment]
@fflorent `log` uses a global `set_logger()` technique: https://github.com/rust-lang-nursery/log/blob/master/src/lib.rs#L1088 which updates the global state so that all subsequent [`log!()`](https://github.com/rust-lang-nursery/log/blob/master/src/macros.rs#L41) calls check it for the actual logging.

**Comment by fflorent** [comment]
Sounds neat! What do you think @colin-kiegel?

Is there any way to set the preferences only once in the tests (for more convenience)?

**Comment by colin-kiegel** [comment]
@fflorent I think, at some point, we could try this. But my priority would be to implement a good default first.

Currently I am working in a local branch with two sets of changes related to this bug
(a) improving the formatting - which is basically done - I am only waiting for more feedback here
(b) a complete re-implementation of the diff-algorithm, which has some weird edge cases like [this issue](https://github.com/johannhof/difference.rs/issues/20). I made good progress today, but it's not fully integrated yet.

**Comment by colin-kiegel** [comment]
Fixing this is blocked by PR https://github.com/johannhof/difference.rs/pull/22 

I decided to at least publish the slightly improved color coding with `v0.4.0`. It should now look like this:

![image](https://user-images.githubusercontent.com/7691186/31588996-44329aa4-b1fa-11e7-95a4-9278dd47497f.png)

Note: The missing line break is still there.

But after the before mentioned PR lands, it will look like this

![image](https://user-images.githubusercontent.com/7691186/31589012-7cbc4186-b1fa-11e7-99b4-2a7132505b9f.png)

**Comment by hcpl** [comment]
@colin-kiegel looks awesome, thanks for your work! I'm not in hurry right now, so I can wait for upstream PR.

**Comment by colin-kiegel** [comment]
@hcpl thx
@fflorent If you want to, feel free to experiment with customizing the output and color coding. I would prefer to keep it simple. Would a feature toggle between variants (A) and (C) be sufficient for you?

**PS**: For the moment I went with (C), because it carries more information, but interestingly in [this poll](https://users.rust-lang.org/t/mega-assert-eq-with-colorful-diff-drop-in-replacement/10101/27), 80% liked variant (A), 28% liked (C) and 12% liked (B). So there would be at least people who prefer (A) over (C). :-)

**Comment by colin-kiegel** [comment]
still waiting for a [new release of `difference.rs`](https://github.com/johannhof/difference.rs/issues/25).

**Comment by colin-kiegel** [comment]
finally fixed with `v0.5.0` :-)
